python/TorchMLIRModule.cpp

```diff
@@ -9,24 +9,49 @@
 
 #include "mlir-c/Bindings/Python/Interop.h"
 #include "mlir/Bindings/Python/PybindAdaptors.h"
+#include "mlir/CAPI/IR.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/InitAllPasses.h"
 #include "torch-mlir-c/Dialects.h"
 #include "torch-mlir-c/Registration.h"
 
 namespace py = pybind11;
 
+using namespace mlir;
+
+void registerUpstreamPasses() {
+  arith::registerArithmeticPasses();
+  bufferization::registerBufferizationPasses();
+  func::registerFuncPasses();
+  registerConvertAffineToStandardPass();
+  registerConvertArithmeticToLLVMPass();
+  registerConvertControlFlowToLLVMPass();
+  registerConvertFuncToLLVMPass();
+  registerConvertLinalgToLLVMPass();
+  registerConvertMathToLLVMPass();
+  registerConvertMemRefToLLVMPass();
+  registerLinalgPasses();
+  registerReconcileUnrealizedCastsPass();
+  registerSCFPasses();
+  registerSCFToControlFlowPass();
+  registerTosaToArithPass();
+  registerTosaToLinalgNamedPass();
+  registerTosaToLinalgPass();
+  tensor::registerTensorPasses();
+}
+
 PYBIND11_MODULE(_torchMlir, m) {
+  registerUpstreamPasses();
   torchMlirRegisterAllPasses();
 
   m.doc() = "torch-mlir main python extension";
 
   m.def(
-      "register_dialect",
-      [](MlirContext context, bool load) {
-        MlirDialectHandle handle = mlirGetDialectHandle__torch__();
-        mlirDialectHandleRegisterDialect(handle, context);
-        if (load) {
-          mlirDialectHandleLoadDialect(handle, context);
-        }
+      "register_required_dialects",
+      [](MlirContext context) {
```

Comment:
Ditto - don't directly use the C++ API here. Until I get all of the dialect handles exposed upstream for individual use, I would recommend creating a torch CAPI (i.e. torchMlirRegisterDialects) that does this. That will keep the layering right and the CAPI self contained.

Reply:
Here is an example of a C API that registers dialects: https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/API/Compiler.cpp#L65